### PR TITLE
Models in r2 rw support

### PR DESCRIPTION
**Pull request by shichengzhou-db**
## Related Issues/PRs

N/A

## What changes are proposed in this pull request?

- Added a R2ArtifactStore class to support registering uc models to be stored in R2 Storage. R2ArtifactStore subclasses S3ArtifactStore to do most of the heavy lifting but overrides some methods to 1) parse (bucket, path) from r2 uri 2) instantiate s3 client with virtual addressing style

## How is this patch tested?

- Added unit test for R2ArtifactStore to specifically test for parsing of R2 URI and if S3 client is initialized correctly for R2 artifact store.
- Manually created a R2 Storage backed UC Catalog and successfully registered a model into it (Version 3 in screenshot).
Screenshot 2023-08-30 at 3 09 06 PM

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Support registering model in a R2 Storage backed Databricks UC Catalog.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [x] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by smurching — APPROVED** [review]
Looks great @shichengzhou-db ! Were you also able to test loading the model that you registered to R2 via e.g. `mlflow.pyfunc.load_model`?

**Comment by smurching** [comment]
@jerrylian-db FYI there were some gotchas we ran into (e.g. needing to use [virtual host style addressing](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html)) to get reading/writing model artifacts to R2 to work using UC temporary credentials. You might want to work with Shicheng to test that R2 model read/write still work once you have a PR ready to refactor S3ArtifactRepo to share the fast upload/download logic in DatabricksArtifactRepo

File: mlflow/store/artifact/r2_artifact_repo.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        host = parsed.netloc
+        path = parsed.path
+
+        bucket = host.split("@")[0]
+        if path.startswith("/"):
+            path = path[1:]
+        return bucket, path
+
+    def _get_s3_client(self, addressing_style="virtual"):
+        return super()._get_s3_client(addressing_style=addressing_style)
```
Comment (smurching) [inline mlflow/store/artifact/r2_artifact_repo.py:26]: You may need to run `black .` to fix lint errors (e.g. having newlines at EOF)

**Comment by shichengzhou-db** [comment]
> Looks great @shichengzhou-db ! Were you also able to test loading the model that you registered to R2 via e.g. `mlflow.pyfunc.load_model`?

yep loaded the model and did a prediction, works as expected (added this manual test code in the notebook shared with you earlier)

**Review by kriscon-db — APPROVED** [review]
LGTM

**Comment by shichengzhou-db** [comment]
@mlflow-automation autoformat

File: mlflow/store/artifact/r2_artifact_repo.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        parsed = urllib.parse.urlparse(uri)
+        if parsed.scheme != "r2":
+            raise Exception(f"Not an R2 URI: {uri}")
+
+        host = parsed.netloc
+        path = parsed.path
+
+        bucket = host.split("@")[0]
+        if path.startswith("/"):
+            path = path[1:]
```
Comment (harupy) [inline mlflow/store/artifact/r2_artifact_repo.py:30]:
```suggestion
        if path.startswith("/"):
            path = path[1:]
```
```suggestion
        path = path.lstrip("/")
```

can we use `lstrip` here?

File: mlflow/store/artifact/r2_artifact_repo.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            access_key_id=access_key_id,
+            secret_access_key=secret_access_key,
+            session_token=session_token,
+        )
+
+    def parse_s3_compliant_uri(self, uri):
+        # r2 uri format(virtual): r2://<bucket-name>@<account-id>.r2.cloudflarestorage.com/<path>
+        parsed = urllib.parse.urlparse(uri)
+        if parsed.scheme != "r2":
+            raise Exception(f"Not an R2 URI: {uri}")
```
Comment (harupy) [inline mlflow/store/artifact/r2_artifact_repo.py:23]:
```suggestion
            raise MlflowException.invalid_parameter_value(f"Not an R2 URI: {uri}")
```

File: mlflow/store/artifact/r2_artifact_repo.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    def __init__(
+        self, artifact_uri, access_key_id=None, secret_access_key=None, session_token=None
+    ):
+        super().__init__(
+            artifact_uri,
+            access_key_id=access_key_id,
+            secret_access_key=secret_access_key,
+            session_token=session_token,
+        )
```
Comment (harupy) [inline mlflow/store/artifact/r2_artifact_repo.py:17]: Do we need to override `__init__`?

**Pull request merged**